lib/licensed/sources/gradle.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+require "tempfile"
+require "csv"
+require "uri"
+require "json"
+require "net/http"
+
+module Licensed
+  module Sources
+    class Gradle < Source
+      def self.type
+        "gradle"
+      end
+
+      def configurations
+        ["runtime", "runtimeClasspath"]
+      end
+
+      def initialize(config)
+        @config = config
+      end
+
+      def enabled?
+        (Licensed::Shell.tool_available?("gradle") || File.executable?(@config.pwd.join("gradlew"))) && File.exist?(@config.pwd.join("build.gradle"))
+      end
+
+      def enumerate_dependencies
+        @dependencies ||= JSON.parse(package_metadata_command).map do |package|
+          name = "#{package["group"]}:#{package["name"]}"
+          Dependency.new(
+            name: name,
+            version: package["version"],
+            path: File.join(@config.pwd, gradle_licenses_path, name),
+            metadata: {
+              "type"     => Gradle.type
+            }
+          )
+        end
+      end
+
+      def with_latest_licenses
+        download_and_cache_licenses do
+          yield
+        end
+      end
+
+      private
+
+      def gradle_licenses_path
+        ".gradle-licenses"
+      end
+
+      def gradle_licenses_csv_name
+        "licenses.csv"
+      end
+
+      def download_licenses_from_csv
+        path = File.join(@config.pwd, gradle_licenses_path)
+        downloaded_licenses = {}
+        CSV.foreach(File.join(path, gradle_licenses_csv_name), headers: true) do |row|
+          url = row["moduleLicenseUrl"]
+          artifact, _, version = row["artifact"].rpartition(":")
+          file_path = File.join(path, artifact, "LICENSE")
+          FileUtils.mkdir_p(File.join(path, artifact))
+          if cache_path = downloaded_licenses[url]
+            FileUtils.cp(cache_path, file_path)
+          else
+            uri = URI(row["moduleLicenseUrl"])
+            File.write(file_path, Net::HTTP.get(uri))
+            downloaded_licenses[url] = file_path
+          end
+        end
+      end
+
+      def download_and_cache_licenses
+        gradle_command("generateLicenseReport")
+        download_licenses_from_csv
+        yield
+        FileUtils.rm_rf(File.join(@config.pwd, gradle_licenses_path))
+      end
+
+      # Returns the output from running `npm list` to get package metadata
+      def package_metadata_command
+        gradle_command("printDependencies")
+      end
+
+      def gradle_command(*args)
+        Tempfile.open(["license", ".gradle"], @config.pwd) do |f|
+          f.write gradle_file
+          f.close
+          Licensed::Shell.execute(executable, "-q", "-b", f.path, *args)
+        end
+      end
+
+      def executable
+        if File.executable?(File.join(@config.pwd, "gradlew"))
+          "./gradlew"
+        else
+          "gradle"
+        end
+      end
+
+      def gradle_file
```

Comment:
This is kinda hacky, but this seemed to be the easiest way to do this and inject additional logic and tasks into the Gradle build process to grab the needed data.

@github/java I'm also curious if you all have better ideas around how to solve this problem to generate license metadata from what Gradle knows about the dependencies.